### ISSUE TITLE
[Suggestion] Make `TensorDict` possibly nested

### ISSUE THREAD

**Issue by vmoens** [facebookresearch/rl#189]
## Description of the problem

TensorDict can currently contain only tensors. However, one could think of having a key pointing to another tensordict:

```python
tensordict1 = TensorDict({'key1': torch.zeros(3, 4), `key2': torch.zeros(3, 4)}, [3, 4])
tensordict2 = TensorDict({'tensordict1': tensordict1, `tensor': torch.zeros(3, 10)}, [3])
```

## Usage

When designing some objects (e.g. losses, multi-step post process etc) we assume that the input is a tensordict. 
The reason for that is that it keeps the training algorithm generic, whatever the name of the relevant keys are: we don't strongly assume that there is one and only one observation, we can pass hidden states if needed, or log-probabilities etc.

However, this heavy reliance on TensorDict classes might slow down adoption of the library: there would be yet another class to learn. Also, it is not clear when looking at a loss module which keys are needed in the tensordict to run (do we need an action log-probability, like in PPO? Do we need a hidden state? etc). Designing losses where the forward method has a specific signature might improve the modularity of the library and make the experiment configuration clearer.

## Proposed refactoring

We should ensure that all methods associated with tensordict work with tensordict values. Since we use some overdubbing of torch functions for tensordicts already I would assume that most of it can be kept unchanged.

We would then need (in a second PR) to refactor a few classes, typically the losses, from
```python
class ThisLossClass(...)
     def forward(self, tensordict: _TensorDict):
         do_fancy_stuff()
```
to this
```python
class ThisLossClass(...)
     def forward(self, observation: Union[torch.Tensor, _TensorDict], next_observation: Union[torch.Tensor, _TensorDict], action: torch.Tensor, done: torch.Tensor, reward: torch.Tensor):
         do_fancy_stuff()
```
In this setting, the observation can be either a tensor or a tensordict, It'll be read by some module in the loss accordingly.

From a training perspective, we'd just need to switch from
```python
loss_td = loss_module(tensordict)
```
to
```python
loss_td = loss_module(**tensordict)
```
provided that the tensordict has been filtered as needed (i.e. we made sure that only the relevant keys persisted).

Perhaps this will break a few things (since the whole loss machinery is currently based upon the assumption that the input is a TensorDict) but on the long run it might improve readability of the code.

cc @shagunsodhani 
cc @Benjamin-eecs
